src/wp-includes/class-wp-query.php

```diff
@@ -3528,8 +3529,20 @@ public function get_posts() {
 			/** @var WP_Post[] */
 			$this->posts = array_map( 'get_post', $this->posts );
 
-			if ( $q['cache_results'] ) {
-				update_post_caches( $this->posts, $post_type, $q['update_post_term_cache'], $q['update_post_meta_cache'] );
+			if ( $q['cache_results'] && $id_posts_are_cacheable ) {
```

Comment:
See my other comment, plus there's one additional thing to consider here: In order to ensure that `$this->posts` contains full post objects, `"{$wpdb->posts}.*" === $fields` must be true, _and_ we have to ensure that none of the previous filters potentially modified the post results. Our source of truth is `$this->posts` in the way that it is set in https://github.com/WordPress/wordpress-develop/blob/3bacfc9f978876549ee0e2b5068be02e006d956c/src/wp-includes/class-wp-query.php#L3337. After that, it may be modified again by certain filters etc., and we cannot assume that the post fields are still intact if any such modifications happened. So I think we should "store" the value from https://github.com/WordPress/wordpress-develop/blob/3bacfc9f978876549ee0e2b5068be02e006d956c/src/wp-includes/class-wp-query.php#L3337 in a variable `$unfiltered_posts` to compare against here. Only if the array still matches, no modifications were made and we can reliably use `$this->posts` to cache the objects.
```suggestion
			if ( $q['cache_results'] && "{$wpdb->posts}.*" === $fields && $unfiltered_posts === $this->posts ) {
```